pennylane/devices/modifiers/convert_single_circuit.py

```diff
@@ -0,0 +1,215 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Defines the ``convert_single_circuit_to_batch`` device modifier.
+
+"""
+from functools import wraps
+
+from pennylane.devices import Device, DefaultExecutionConfig
+from pennylane.tape import QuantumScript
+
+
+def _make_execute(batch_execute):
+    """Allows an ``execute`` function to handle individual circuits."""
+
+    @wraps(batch_execute)
+    def execute(self, circuits, execution_config=DefaultExecutionConfig):
```

Comment:
I'm curious, why do we allow single circuits to be passed to a devices `execute` method?

Reply:
While the single-circuit behavior isn't actually used by the workflow, it can be convenient for messing around with the devices themselves. I think this was motivated the previous existence of both `qml.Device.execute` and `qml.Device.batch_execute`.

Reply:
> While the single-circuit behavior isn't actually used by the workflow, it can be convenient for messing around with the devices themselves.

@albi3ro curious what problem we're aiming to solve in this case 🤔

Reply:
So it looks like all the conversions we had around this last year were verbal. I just read over all the discussions on PR #3602  and nothing seemed to be written down. This may be another example of historical thinking (the ability to execute both a single circuit and a batch) being hard to break out of.

Reply:
Note that we could also create two other device modifiers `add_multiprocessing` and `iterate_over_batches` (or some other names).  In those cases, the developer can just define `Device.execute` that accepts a single circuit, and then we just iterate over that call.

```
@iterate_over_batches
class MyDevice(qml.devices.Device):

    def execute(self, circuits: QuantumTape, execution_config = qml.devices.DefaultExecutionConfig):
        return simulate_single_circuit(circuit)
```

In that case, the ability to just define the single circuit execution will be really handy.

Reply:
Could having `execute` accept a single circuit + having the `convert_single_circuit_to_batch` decorator help making it easier to port over the devices from the old API?

Reply:
Potentially. It comes back to whether the devices has special logic for handling batches or not.